open_spiel/games/battleship.cc

```diff
@@ -0,0 +1,849 @@
+      if (next_ship.length > 1 && next_ship.length <= conf.board_height) {
+        for (int row = 0; row < conf.board_height - next_ship.length + 1;
+             ++row) {
+          for (int col = 0; col < conf.board_width - next_ship.length; ++col) {
+            const ShipPlacement placement(ShipPlacement::Direction::Vertical,
+                                          /* ship = */ next_ship,
+                                          /* tl_corner = */ Cell{row, col});
+            if (PlacementDoesNotOverlap(placement, player)) {
+              actions.push_back(SerializeShipPlacementAction(placement));
+            }
+          }
+        }
+      }
+
+      // FIXME(gfarina): It would be better to have a check of this time at game
+      //     construction time.
+      if (actions.empty()) {
+        SpielFatalError(
+            "Battleship: it is NOT possible to fit all the ships on the "
+            "board!");
+      }
+    } else {
+      // In this case, the only thing the player can do is to shoot on a cell
+      //
+      // Depending on whether repeated shots are allowed or not, we might filter
+      // out some cells.
+      for (int row = 0; row < conf.board_height; ++row) {
+        for (int col = 0; col < conf.board_width; ++col) {
+          if (!conf.allow_repeated_shots &&
+              AlreadyShot(Cell{row, col}, CurrentPlayer())) {
+            // We do not duplicate the shot, so nothing to do here...
+          } else {
+            actions.push_back(SerializeShotAction(Shot{row, col}));
+          }
+        }
+      }
+
+      // SAFETY: The assert below can never fail, because when
+      //     allow_repeated_shot is false, we check at game construction time
+      //     that the number of shots per player is <= the number of cells in
+      //     the board.
+      SPIEL_DCHECK_FALSE(actions.empty());
+    }
+
+    return actions;
+  }
+}
+
+std::string BattleshipState::ActionToString(Player player,
+                                            Action action) const {
+  SPIEL_DCHECK_TRUE(player == Player1 || player == Player2);
+
+  // FIXME(gfarina): It was not clear to me from the documentation whether
+  //     the next condition is always guaranteed.
+  //
+  //     Action ids are reused at different states, so the meaning of the same
+  //     action id is contingent on the current state.
+  SPIEL_CHECK_EQ(player, CurrentPlayer());
+
+  if (!AllShipsPlaced()) {
+    // If we are here, we still have some ships to place on the board.
+    //
+    // First, we find the first ship that hasn't been placed on the board yet.
+    const ShipPlacement ship_placement = DeserializeShipPlacementAction(action);
+    return ship_placement.ToString();
+  } else {
+    // In this case, the only thing the player can do is to shoot on a cell
+    const Shot shot = DeserializeShotAction(action);
+    return shot.ToString();
+  }
+}
+
+std::string BattleshipState::ToString() const {
+  std::string state_str;
```

Comment:
Not sure if applicable, but could you make it into two 2D grids, something like . for water, X for battleship, o for missed, * for destroyed part of battleship. It would be visually meaningful at first glance.

Reply:
Ah that's a good idea! :slightly_smiling_face: The only worry is that it would create way longer state strings, and if implemented algorithms keep internal maps keyed on state strings, that could lead to performance issues.

Either way, I think the feature you mentioned makes sense to have, so I will implement it. If `ToString` turns out to be important for performance, we could have what you were asking under a different name, like `ToPrettyString` or `BoardStrings` or something like that.

Reply:
Done, I've added three additional public methods that print the boards. I've left a `FIXME` in the code about whether it would make sense to have `ToStateString` be replaced by `ToPrettyString`. I'll let you and @lanctot figure that out, since I don't have a good understanding of how `ToString` is used in practice by the different algorithms.

Reply:
Cool, but please change it to TODO or I won't be able to import it.